File: src/main/cpp/utils/profiler.cpp
/*! 
 * \author gold1 
 * \date 30-Aug-17.
 */

#include <cstring>
#include "profiler.h"
#include <easylogging++.h>

namespace nova {
    std::unordered_map<std::string, profiler_data> profiler::data;

    void profiler::start(std::string name) {
        auto section_itr = data.find(name);
        if(section_itr == data.end()) {
            data[name] = profiler_data();
        }

        auto &cur_profiler_data = data[name];
        cur_profiler_data.start_time = std::chrono::high_resolution_clock::now();
    }

    void profiler::end(std::string name) {
        auto &cur_profiler_data = data[name];
        auto duration = std::chrono::high_resolution_clock::now() - cur_profiler_data.start_time;
        cur_profiler_data.total_duration += duration;
    }

    void profiler::log_all_profiler_data() {
        std::stringstream ss;
        for(const auto& item : data) {
            const auto& cur_profiler_data = item.second;

            ss << "Profiled section " << item.first << " has taken an total of " << double(std::chrono::duration_cast<std::chrono::nanoseconds>(cur_profiler_data.total_duration).count()) / 1000000.0f << "ms to execute since the game began\n";
        }

        //LOG_EVERY_N(100, DEBUG) << ss.str();
    }
}

File: src/main/java/com/continuum/nova/utils/Profiler.java
package com.continuum.nova.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.HashMap;
import java.util.Map;

/**
 * Provides a simpler profiling interface than Minecraft's profiler
 *
 * @author ddubois
 * @since 31-Aug-17
 */
public class Profiler {
    private static final Logger LOG = LogManager.getLogger(Profiler.class);

    public static final int NUM_SAMPLES = 120;

    public static class ProfilerData {
        long startTime;
        long total_duration = 0;

        public void start() {
            startTime = System.nanoTime();
        }

        public void stop() {
            long endTime = System.nanoTime();
            total_duration += endTime - startTime;
        }
    }

    private static Map<String, ProfilerData> profilerDataMap = new HashMap<>();

    public static void start(String name) {
        if(!profilerDataMap.containsKey(name)) {
            profilerDataMap.put(name, new ProfilerData());
        }

        profilerDataMap.get(name).start();
    }

    public static void end(String name) {
        profilerDataMap.get(name).stop();
    }

    private static int counter = 0;

    public static void logData() {
        if(counter >= 100) {
            for(Map.Entry<String, ProfilerData> entry : profilerDataMap.entrySet()) {
                LOG.debug("Section {} has taken an total of {}ms since the game began", entry.getKey(), (double) entry.getValue().total_duration / 1000000.0);
            }
            counter = 0;
        }
        counter++;
    }

    private Profiler() {}
}
